lib/workflow/postgres/sql_schema.go

```diff
@@ -0,0 +1,309 @@
+    workflow_instance_id BIGINT,
+    enqueued_at TIMESTAMPTZ NOT NULL DEFAULT NOW(),
+    -- task_complete members
+    task_result_id BIGINT
+);
+
+-- Workflow Functions
+CREATE OR REPLACE FUNCTION enqueue_workflow(
+    name TEXT,
+    workflow_name TEXT,
+    parameters BYTEA)
+RETURNS INTEGER
+AS $$
+    WITH winst AS (
+        INSERT INTO workflow_instances(name, workflow_name, parameters)
+            VALUES(name, workflow_name, parameters)
+            ON CONFLICT DO NOTHING
+            RETURNING id
+        )
+    INSERT INTO workflow_events(event_type, workflow_instance_id)
+    (SELECT 'start', id FROM winst WHERE id IS NOT NULL)
+    RETURNING 1
+$$ LANGUAGE SQL;
+
+CREATE OR REPLACE FUNCTION dequeue_workflow(VARIADIC workflow_names TEXT[])
+RETURNS TABLE(workflow_instance_id BIGINT, instance_name TEXT, workflow_name TEXT,
+    status workflow_instance_status, parameters BYTEA, payload BYTEA, event_id BIGINT,
+    event_type workflow_event_type, task_result_id BIGINT, enqueued_tasks INTEGER,
+    completed_tasks INTEGER)
+AS $$
+    WITH nextwinst AS (
+        SELECT
+            a.id id,
+            a.name instance_name,
+            a.workflow_name workflow_name,
+            a.status status,
+            a.parameters parameters,
+            a.payload payload,
+            b.id event_id,
+            b.event_type event_type,
+            b.task_result_id task_result_id,
+            a.enqueued_tasks,
+            a.completed_tasks
+        FROM workflow_instances a
+        INNER JOIN workflow_events b ON a.id = b.workflow_instance_id
+        WHERE a.workflow_name = ANY(workflow_names)
+        ORDER BY b.enqueued_at FOR UPDATE SKIP LOCKED LIMIT 1
+    ),
+    updated AS (
+        UPDATE workflow_instances w1 SET updated_at = NOW()
+        WHERE w1.id = (
+            SELECT id FROM nextwinst
+        )
+    )
+    SELECT * from nextwinst
+$$ LANGUAGE SQL;
+
+CREATE OR REPLACE FUNCTION cancel_workflow(workflow_instance_id BIGINT)
+RETURNS VOID
+AS $$
+    INSERT INTO workflow_events(event_type, workflow_instance_id)
+        VALUES('cancel', workflow_instance_id);
+    -- TODO: notify
+$$ LANGUAGE SQL;
+
+CREATE OR REPLACE FUNCTION abandon_workflow(_workflow_instance_id BIGINT, eid BIGINT, _completed_tasks INTEGER)
+RETURNS VOID
+AS $$
+    WITH unclaimed_tasks AS (
+        SELECT id FROM tasks
+        WHERE workflow_instance_id = _workflow_instance_id
+        FOR UPDATE SKIP LOCKED
+    ), deleted_tasks AS (
+        DELETE FROM tasks WHERE id IN (SELECT id from unclaimed_tasks)
+    )
+    UPDATE workflow_instances w1 SET updated_at = NOW(), status = 'abandoned',
+        completed_tasks = _completed_tasks + (select count(*) from unclaimed_tasks);
+
+    DELETE FROM workflow_events WHERE id = eid;
+
+    INSERT INTO workflow_events(event_type, workflow_instance_id)
+        VALUES('tasks_abandoned', _workflow_instance_id);
+$$ LANGUAGE SQL;
+
+
+CREATE OR REPLACE FUNCTION cancel_workflow(workflow_instance_id BIGINT)
```

Comment:
This is a redefinition of the function at line 197 but is wrong, let's remove it!